### ISSUE TITLE
`@task` and `@workflow` decorators bring a lot of overhead during code execution.

### ISSUE THREAD

**Issue by anmyachev** [flyteorg/flyte#2113]
With decorators, the execution time is about 50 seconds longer than without.

What internal `flyte` mechanisms can bring such an overhead? Can it be reduced in some way by the user? Is it possible to optimize in developers side?

flytekit version: `0.26.0`

<details><summary>Code</summary>

```python
import time

# from collections import OrderedDict
# from dataclasses import dataclass
import typing
import warnings
from timeit import default_timer as timer

# from flytekit.types.schema import FlyteSchema
import numpy as np
import pandas as pd
import sklearn.linear_model as lm
from flytekit import task, workflow

# from flytekit import Resources
from flytekit.types.file import FlyteFile
from sklearn import config_context
from sklearn.model_selection import train_test_split

warnings.filterwarnings("ignore")

ML_KEYS = ["t_train_test_split", "t_train", "t_inference", "t_ml"]
ML_SCORE_KEYS = ["mse_mean", "cod_mean", "mse_dev"]

N_RUNS = 50
TEST_SIZE = 0.1
RANDOM_STATE = 777

# Functions from utils


def mse(y_test, y_pred):
    return ((y_test - y_pred) ** 2).mean()


def cod(y_test, y_pred):
    y_bar = y_test.mean()
    total = ((y_test - y_bar) ** 2).sum()
    residuals = ((y_test - y_pred) ** 2).sum()
    return 1 - (residuals / total)


def split(X, y, test_size=0.1, stratify=None, random_state=None):
    t0 = timer()
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=test_size, stratify=stratify, random_state=random_state
    )
    split_time = timer() - t0

    return (X_train, y_train, X_test, y_test), split_time


COLS = [
    "YEAR",
    "DATANUM",
    "SERIAL",
    "CBSERIAL",
    "HHWT",
    "CPI99",
    "GQ",
    "PERNUM",
    "SEX",
    "AGE",
    "INCTOT",
    "EDUC",
    "EDUCD",
    "EDUC_HEAD",
    "EDUC_POP",
    "EDUC_MOM",
    "EDUCD_MOM2",
    "EDUCD_POP2",
    "INCTOT_MOM",
    "INCTOT_POP",
    "INCTOT_MOM2",
    "INCTOT_POP2",
    "INCTOT_HEAD",
    "SEX_HEAD",
]

COLUMNS_TYPES = [
    "int",
    "int",
    "int",
    "float",
    "int",
    "float",
    "int",
    "float",
    "int",
    "int",
    "int",
    "int",
    "int",
    "int",
    "int",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
]

# X = OrderedDict((zip(cols, column_types)))
# Y = OrderedDict({"EDUC": X.pop("EDUC")})


@task
def feature_eng_task(
    data: FlyteFile[typing.TypeVar("csv")], cols: typing.List[str]
) -> (pd.DataFrame):

    df = pd.read_csv(data, compression="infer")[cols]

    df = df[df["INCTOT"] != 9999999]
    df = df[df["EDUC"] != -1]
    df = df[df["EDUCD"] != -1]

    df["INCTOT"] = df["INCTOT"] * df["CPI99"]

    for column in cols:
        df[column] = df[column].fillna(-1)
        df[column] = df[column].astype("float64")

    return df


@task
def ml_task(
    df: pd.DataFrame,
    random_state: int,
    n_runs: int,
    test_size: float,
    ml_keys: typing.List[str],
    ml_score_keys: typing.List[str],
) -> (typing.Dict[str, float], typing.Dict[str, float]):

    # Fetch the input and output data from train dataset
    y = np.ascontiguousarray(df["EDUC"], dtype=np.float64)
    X = np.ascontiguousarray(df.drop(columns=["EDUC", "CPI99"]), dtype=np.float64)

    clf = lm.Ridge()

    mse_values, cod_values = [], []
    ml_times = {key: 0.0 for key in ml_keys}
    ml_scores = {key: 0.0 for key in ml_score_keys}

    print("ML runs: ", n_runs)
    for i in range(n_runs):
        (X_train, y_train, X_test, y_test), split_time = split(
            X, y, test_size=test_size, random_state=random_state
        )
        ml_times["t_train_test_split"] += split_time
        random_state += 777

        t0 = timer()
        with config_context(assume_finite=True):
            model = clf.fit(X_train, y_train)
        ml_times["t_train"] += timer() - t0

        t0 = timer()
        y_pred = model.predict(X_test)
        ml_times["t_inference"] += timer() - t0

        mse_values.append(mse(y_test, y_pred))
        cod_values.append(cod(y_test, y_pred))

    ml_times["t_ml"] += ml_times["t_train"] + ml_times["t_inference"]

    ml_scores["mse_mean"] = sum(mse_values) / len(mse_values)
    ml_scores["cod_mean"] = sum(cod_values) / len(cod_values)
    ml_scores["mse_dev"] = pow(
        sum([(mse_value - ml_scores["mse_mean"]) ** 2 for mse_value in mse_values])
        / (len(mse_values) - 1),
        0.5,
    )
    ml_scores["cod_dev"] = pow(
        sum([(cod_value - ml_scores["cod_mean"]) ** 2 for cod_value in cod_values])
        / (len(cod_values) - 1),
        0.5,
    )

    return ml_scores, ml_times


@workflow
def census_bench_wf(
    dataset: FlyteFile[
        "csv"  # noqa F821
    ] = "https://rapidsai-data.s3.us-east-2.amazonaws.com/datasets/ipums_education2income_1970-2010.csv.gz",
    cols: typing.List[str] = COLS,
    random_state: int = RANDOM_STATE,
    n_runs: int = N_RUNS,
    test_size: float = TEST_SIZE,
    ml_keys: typing.List[str] = ML_KEYS,
    ml_score_keys: typing.List[str] = ML_SCORE_KEYS,
) -> (typing.Dict[str, float], typing.Dict[str, float]):
    df = feature_eng_task(data=dataset, cols=cols)
    ml_scores, ml_times = ml_task(
        df=df,
        random_state=random_state,
        n_runs=n_runs,
        test_size=test_size,
        ml_keys=ml_keys,
        ml_score_keys=ml_score_keys,
    )
    return ml_scores, ml_times


if __name__ == "__main__":
    start = time.time()
    print(census_bench_wf())
    print("--- %s seconds ---" % (time.time() - start))

```

</details>

**Comment by kumare3** [comment]
@anmyachev when you run it locally here are few things to note about the local scheduler
1. It is a linear scheduler - i.e. does not automatically parallelize based on the dataflow. When running on the cloud scheduler or in sandbox, the execution is auto-parallelized
2. There is also overhead between consecutive task executions, as we are trying to mimic the behavior of the actual scheduler and try to help identify bugs / issues.
3. 50 seconds still feel longer. 
cc @eapolinario / @wild-endeavor

**Comment by eapolinario** [comment]
@anmyachev , can you provide a few more details? For example, are you running this locally or in a VM in the cloud? How did you measure the overhead? How many times did you run this code?

Answering some of these questions will help us understand the source of the overhead. For example, the code you provided depends on a 360MB file, that alone can skew the execution time tremendously. 

I ran this modified version of the script you provided and noticed that running this locally the huge variance comes from downloading the dataset.

Here's the modified version of the script:

<details><summary>Code</summary>

```python
import time

# from collections import OrderedDict
# from dataclasses import dataclass
import typing
import warnings
from timeit import default_timer as timer

# from flytekit.types.schema import FlyteSchema
import numpy as np
import pandas as pd
import sklearn.linear_model as lm
from flytekit import task, workflow

# from flytekit import Resources
from flytekit.types.file import FlyteFile
from sklearn import config_context
from sklearn.model_selection import train_test_split

warnings.filterwarnings("ignore")

ML_KEYS = ["t_train_test_split", "t_train", "t_inference", "t_ml"]
ML_SCORE_KEYS = ["mse_mean", "cod_mean", "mse_dev"]

N_RUNS = 50
TEST_SIZE = 0.1
RANDOM_STATE = 777

# Functions from utils


def mse(y_test, y_pred):
    return ((y_test - y_pred) ** 2).mean()


def cod(y_test, y_pred):
    y_bar = y_test.mean()
    total = ((y_test - y_bar) ** 2).sum()
    residuals = ((y_test - y_pred) ** 2).sum()
    return 1 - (residuals / total)


def split(X, y, test_size=0.1, stratify=None, random_state=None):
    t0 = timer()
    X_train, X_test, y_train, y_test = train_test_split(
        X, y, test_size=test_size, stratify=stratify, random_state=random_state
    )
    split_time = timer() - t0

    return (X_train, y_train, X_test, y_test), split_time


COLS = [
    "YEAR",
    "DATANUM",
    "SERIAL",
    "CBSERIAL",
    "HHWT",
    "CPI99",
    "GQ",
    "PERNUM",
    "SEX",
    "AGE",
    "INCTOT",
    "EDUC",
    "EDUCD",
    "EDUC_HEAD",
    "EDUC_POP",
    "EDUC_MOM",
    "EDUCD_MOM2",
    "EDUCD_POP2",
    "INCTOT_MOM",
    "INCTOT_POP",
    "INCTOT_MOM2",
    "INCTOT_POP2",
    "INCTOT_HEAD",
    "SEX_HEAD",
]

COLUMNS_TYPES = [
    "int",
    "int",
    "int",
    "float",
    "int",
    "float",
    "int",
    "float",
    "int",
    "int",
    "int",
    "int",
    "int",
    "int",
    "int",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
    "float",
]

# X = OrderedDict((zip(cols, column_types)))
# Y = OrderedDict({"EDUC": X.pop("EDUC")})


@task
def feature_eng_task(
    data: FlyteFile[typing.TypeVar("csv")], cols: typing.List[str]
) -> (pd.DataFrame, float):
    start = time.time()
    df = pd.read_csv(data, compression="infer")[cols]
    download_total_time = time.time() - start

    df = df[df["INCTOT"] != 9999999]
    df = df[df["EDUC"] != -1]
    df = df[df["EDUCD"] != -1]

    df["INCTOT"] = df["INCTOT"] * df["CPI99"]

    for column in cols:
        df[column] = df[column].fillna(-1)
        df[column] = df[column].astype("float64")

    return df, download_total_time


@task
def ml_task(
    df: pd.DataFrame,
    random_state: int,
    n_runs: int,
    test_size: float,
    ml_keys: typing.List[str],
    ml_score_keys: typing.List[str],
) -> (typing.Dict[str, float], typing.Dict[str, float]):
    # Fetch the input and output data from train dataset
    y = np.ascontiguousarray(df["EDUC"], dtype=np.float64)
    X = np.ascontiguousarray(df.drop(columns=["EDUC", "CPI99"]), dtype=np.float64)

    clf = lm.Ridge()

    mse_values, cod_values = [], []
    ml_times = {key: 0.0 for key in ml_keys}
    ml_scores = {key: 0.0 for key in ml_score_keys}

    print("ML runs: ", n_runs)
    for i in range(n_runs):
        (X_train, y_train, X_test, y_test), split_time = split(
            X, y, test_size=test_size, random_state=random_state
        )
        ml_times["t_train_test_split"] += split_time
        random_state += 777

        t0 = timer()
        with config_context(assume_finite=True):
            model = clf.fit(X_train, y_train)
        ml_times["t_train"] += timer() - t0

        t0 = timer()
        y_pred = model.predict(X_test)
        ml_times["t_inference"] += timer() - t0

        mse_values.append(mse(y_test, y_pred))
        cod_values.append(cod(y_test, y_pred))

    ml_times["t_ml"] += ml_times["t_train"] + ml_times["t_inference"]

    ml_scores["mse_mean"] = sum(mse_values) / len(mse_values)
    ml_scores["cod_mean"] = sum(cod_values) / len(cod_values)
    ml_scores["mse_dev"] = pow(
        sum([(mse_value - ml_scores["mse_mean"]) ** 2 for mse_value in mse_values])
        / (len(mse_values) - 1),
        0.5,
    )
    ml_scores["cod_dev"] = pow(
        sum([(cod_value - ml_scores["cod_mean"]) ** 2 for cod_value in cod_values])
        / (len(cod_values) - 1),
        0.5,
    )

    return ml_scores, ml_times


@workflow
def census_bench_wf(
    dataset: FlyteFile[
        "csv"  # noqa F821
    ] = "https://rapidsai-data.s3.us-east-2.amazonaws.com/datasets/ipums_education2income_1970-2010.csv.gz",
    cols: typing.List[str] = COLS,
    random_state: int = RANDOM_STATE,
    n_runs: int = N_RUNS,
    test_size: float = TEST_SIZE,
    ml_keys: typing.List[str] = ML_KEYS,
    ml_score_keys: typing.List[str] = ML_SCORE_KEYS,
) -> (typing.Dict[str, float], typing.Dict[str, float], float):
    df, download_dataset_time = feature_eng_task(data=dataset, cols=cols)
    ml_scores, ml_times = ml_task(
        df=df,
        random_state=random_state,
        n_runs=n_runs,
        test_size=test_size,
        ml_keys=ml_keys,
        ml_score_keys=ml_score_keys,
    )
    return ml_scores, ml_times, download_dataset_time


if __name__ == "__main__":
    start = time.time()
    ml_scores, ml_times, download_dataset_total_time = census_bench_wf()
    print(f"ml_scores={ml_scores}, ml_times={ml_times}")

    total_time = time.time() - start
    total_time_minus_download_dataset = total_time - download_dataset_total_time
    print(f"--- total time in seconds is {total_time}")
    print(f"--- total time minus time to download datset in seconds is {total_time_minus_download_dataset}")

```

</details>


And the result of a few runs:
```
Attempt #1 
❯ python overhead.py
ML runs:  50
ml_scores={'mse_mean': 0.03256456908804958, 'cod_mean': 0.9953675334603812, 'mse_dev': 4.1799404199947226e-05, 'cod_dev': 5.86922791217315e-06}, ml_times={'t_train_test_split': 79.20308212998862, 't_train': 85.51357001600627, 't_inference': 1.5760020540037658, 't_ml': 87.08957207001004}
--- total time in seconds is 342.3200669288635
--- total time minus time to download datset in seconds is 178.59921646118164

Attempt #2
❯ python overhead.py
ML runs:  50
ml_scores={'mse_mean': 0.03256456908804958, 'cod_mean': 0.9953675334603812, 'mse_dev': 4.1799404199947226e-05, 'cod_dev': 5.86922791217315e-06}, ml_times={'t_train_test_split': 79.07691173201601, 't_train': 83.07979319199876, 't_inference': 1.453254410003865, 't_ml': 84.53304760200263}
--- total time in seconds is 316.51622772216797
--- total time minus time to download datset in seconds is 176.09522366523743

Attempt #3
❯ python overhead.py
ML runs:  50
ml_scores={'mse_mean': 0.03256456908804958, 'cod_mean': 0.9953675334603812, 'mse_dev': 4.1799404199947226e-05, 'cod_dev': 5.86922791217315e-06}, ml_times={'t_train_test_split': 78.55288129199835, 't_train': 84.27626563701051, 't_inference': 1.5048866800243559, 't_ml': 85.78115231703487}
--- total time in seconds is 319.0885615348816
--- total time minus time to download datset in seconds is 175.893794298172
```

Notice how the download times vary wildly between the runs while the total execution time is relatively close to each other.

**Comment by eapolinario** [comment]
@anmyachev , I'm closing this issue, but please feel free to reopen if you need more clarification.
